lightning/src/ln/onion_utils.rs

```diff
@@ -436,20 +442,28 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 									is_permanent: true,
 								});
 							}
-						} else if !payment_failed {
+						} else if payment_failed {
+							// Only blame the hop when a value in the HTLC doesn't match the
+							// corresponding value in the onion.
+							short_channel_id = match error_code & 0xff {
+								18|19 => Some(route_hop.short_channel_id),
+								_ => None,
+							};
+						} else {
 							// We can't understand their error messages and they failed to
 							// forward...they probably can't understand our forwards so its
 							// really not worth trying any further.
 							network_update = Some(NetworkUpdate::NodeFailure {
 								node_id: route_hop.pubkey,
 								is_permanent: true,
```

Comment:
It does seem odd to generate a permanent node failure without banning all of that node's channels. Should there be a todo?

Reply:
Now that I think of it, processing this `NetworkUpdate` should result in the node and all its channels being removed from `NetworkGraph` when the event is handled by `NetGraphMsgHandler`.

Reply:
Actually, I'm mistaken here. This happens through `NetworkGraph`'s interface but our code for processing `NodeFailure` just has a TODO already.